WalletWasabi/Services/WalletService.cs

```diff
@@ -1068,6 +1073,68 @@ public IEnumerable<string> GetNonSpecialLabels()
 				.Distinct();
 		}
 
+		private async Task RefreshHistoryAsync(SemaphoreSlim semaphore, SmartCoin coin, ILookup<Script, SmartCoin> lookupScriptPubKey, ILookup<uint256, SmartCoin> lookupSpenderTransactionId, ILookup<uint256, SmartCoin> lookupTransactionId)
+		{
+			if (semaphore != null)
+			{
+				await semaphore.WaitAsync();
+			}
+			var result = string.Join(", ", GetHistory(coin, new List<SmartCoin>(), lookupScriptPubKey, lookupSpenderTransactionId, lookupTransactionId).Select(x => x.Label).Distinct());
+			coin.SetHistory(result);
+			semaphore?.Release();
+		}
+
+		private volatile int _refreshCoinCalls;
+
+		public async void RefreshCoinsHistoriesAsync()
+		{
+			try
+			{
+				if (_refreshCoinCalls == 2) //it is running and scheduled to rerun after finished
+				{
+					return;
+				}
+				if (_refreshCoinCalls == 1) //it is running but now we will rerun if finished
+				{
+					Interlocked.Increment(ref _refreshCoinCalls);
+					return;
+				}
+				if (_refreshCoinCalls == 0) //it is not running so we start the work
+				{
+					Interlocked.Increment(ref _refreshCoinCalls);
+				}
+				var unspentCoins = Coins.Where(c => c.Unspent); //refreshing unspent coins history only
+				if (unspentCoins.Any())
+				{
+					ILookup<Script, SmartCoin> lookupScriptPubKey = Coins.ToLookup(c => c.ScriptPubKey, c => c);
+					ILookup<uint256, SmartCoin> lookupSpenderTransactionId = Coins.ToLookup(c => c.SpenderTransactionId, c => c);
+					ILookup<uint256, SmartCoin> lookupTransactionId = Coins.ToLookup(c => c.TransactionId, c => c);
+
+					const int simultaneousThread = 2; //threads allowed to run simultaneously in threadpool
+					using (SemaphoreSlim semaphore = new SemaphoreSlim(simultaneousThread))
+					{
+						//https://blogs.msdn.microsoft.com/andrewarnottms/2017/05/11/limiting-concurrency-for-faster-and-more-responsive-apps/
+						IEnumerable<Task> tasks = unspentCoins.Select(c => Task.Run(() => RefreshHistoryAsync(semaphore, c, lookupScriptPubKey, lookupSpenderTransactionId, lookupTransactionId)));
+						await Task.WhenAll(tasks); //await all tasks to finish
```

Comment:
Isn't `Parallel.ForEach`, since it's CPU bound, not IO bound.

Reply:
57 coin in the list with ~22 history per coin
**With Parallel.ForEach** => 5091 ms
```
await Task.Run(() => Parallel.ForEach(unspentCoins, new ParallelOptions { MaxDegreeOfParallelism = simultaneousThread }, coin =>
{
	var result = string.Join(", ", GetHistory(coin, new List<SmartCoin>(), lookupScriptPubKey, lookupSpenderTransactionId, lookupTransactionId).Select(x => x.Label).Distinct());
	coin.SetHistory(result);
}));
```
Only two task at once in TaskPool
![image](https://user-images.githubusercontent.com/9844978/51246855-b1769d00-198b-11e9-8199-78b0c8fba2d8.png)

**With Task.WhenAll** => 4592 ms

```
TaskScheduler scheduler = new ConcurrentExclusiveSchedulerPair(TaskScheduler.Default, simultaneousThread).ConcurrentScheduler;

IEnumerable<Task> tasks = unspentCoins.Select(coin => Task.Factory.StartNew(() =>
					   {
						   var result = string.Join(", ", GetHistory(coin, new List<SmartCoin>(), lookupScriptPubKey, lookupSpenderTransactionId, lookupTransactionId).Select(x => x.Label).Distinct());
						   coin.SetHistory(result);
					   }
						, CancellationToken.None, TaskCreationOptions.None, scheduler));
					await Task.WhenAll(tasks); //await all tasks to finish

```

![image](https://user-images.githubusercontent.com/9844978/51247030-2a75f480-198c-11e9-9338-e5eb41ef5702.png)

**Conclusion**

I prefer Parallel.Foreach. As you can see when I use Task.WhenAll the ThreadPool is filled with "Scheduled and waiting to run" tasks and I would like to keep that number low despite of that the Parallel.Foreach is slower (I think because of the task creation overhead).